text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,485 @@
+
+## Lifetime elision
+
+Arbitrary `self` parameters may involve lifetimes.
+
+Even in existing stable Rust, there are [bugs in lifetime elision for complex `Self` types such as `&Box<Self>`](https://github.com/rust-lang/rust/issues/117715). We're aiming to fix them whether or not this RFC is accepted. The net rules will be:
+
+* If a parameter is the first parameter, and
+* Called `self`, and
+* Its type involves `Self` anywhere, and
+* Its type contains _exactly one_ lifetime anywhere
+
+then that lifetime may be used to elide lifetimes on return types, and will take precedence over any lifetimes in other parameters.
+
+If this seems wrong, please discuss this over on [the linked bug](https://github.com/rust-lang/rust/issues/117715) rather than here in this RFC, because none of that should change with this RFC (though it does make it more likely users will run into the current inconsistencies). We'll try to keep this RFC up to date with the outcome of those discussions.
+
+## Diagnostics
+[diagnostics]: #diagnostics
+
+The existing branches in the compiler for "arbitrary self types" already emit excellent diagnostics. We will largely re-use them, with the following improvements:
+
+- In the case where a self type is invalid because it doesn't implement `Receiver`, the existing excellent error message will be updated.
+- An easy mistake is to implement `Receiver` for `P<T>`, forgetting to specify `T: ?Sized`. `P<Self>` then only works as a `self` parameter in traits `where Self: Sized`, an unusual stipulation. It's not obvious that `Sized`ness is the problem here, so we will identify this case specifically and produce an error giving that hint.
+- There are certain types which feel like they "should" implement `Receiver` but do not: `Weak` and `NotNull`. If these are encountered as a self type, we should produce a specific diagnostic explaining that they do not implement `Receiver` and suggesting that they could be wrapped in a newtype wrapper if method calls are important. We hope this can be achieved with [diagnostic items](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html).
+- The current unstable arbitrary self types feature allows generic receivers. For instance,
+  ```rust
+  impl Foo {
+    fn a<R: Deref<Target=Self>>(self: R) { }
+  }
+  ```
+  We don't know a use-case for this. There are several cases where this can result in misleading diagnostics. (For instance, if such a method is called with an incorrect type (for example `smart_ptr.a::<&Foo>()` instead of `smart_ptr.a::<Foo>()`). We could attempt to find and fix all those cases. However, we feel that generic receiver types might risk subtle interactions with method resolutions and other parts of the language. We think it is a safer choice to generate an error on any declaration of a generic `self` type.
```

Comment:
How broad is this? We have one where case where we want a smart pointer with two generic parameters. For example:
```rs
struct ListArc<T: ?Sized, const ID: u64> {
    ptr: NonNull<ArcInner<T>>,
}

impl MyStruct {
    fn method1(self: ListArc<Self, 17>) {}
    fn method2<const ID: u64>(self: ListArc<Self, ID>) {}
}
```
Would `method1` be allowed? Would `method2` be allowed? What if `ID` was a type rather than a const generic?

Reply:
That's fine because the concrete type is known